### ISSUE TITLE
Discussion: Rails/HelperInstanceVariable

### ISSUE THREAD

**Issue by FLeinzi** [makandra/makandra-rubocop#12]
https://www.rubydoc.info/gems/rubocop-rails/RuboCop/Cop/Rails/HelperInstanceVariable

Example:
```
app/helpers/some_helper.rb:5:34: C: Rails/HelperInstanceVariable: Do not use instance variables in helpers.
  return if Rails.env.test? && @instance_variable
```
I want to disable this cop since we sometimes have used instance variables in helpers for good reasons.

**Comment by foobear** [comment]
I agree that there are reasons. Voting to disable this cop = :-1:

**Comment by mak-dunkelziffer** [comment]
I just had a 90 minute debugging session due to this rule being disabled.

I don't believe that there are good reasons to access a "global" instance variable from a global helper. In which situation wouldn't it work to pass in the instance variable as an argument to decouple the code?

I generally prefer "locality of behavior". I don't like views being a huge ball-of-mud were you can't even do dead-code elimination anymore.

**Comment by foobear** [comment]
True, it usually is a smell of bad implementation. I myself don't want to see helper methods or view partials expecting instance variables to exist.

However, If it's done accidentally, it's usually not the end of the world, and I'm sure there are (rare) valid use cases for doing so intentionally. Probably more so in legacy projects from times before global stores like `Current` .

I'm not sure if one bad debugging session makes a valid case for enabling a cop that likely causes offenses in existing projects. Especially since there is no silver bullet to fixing those.

**Comment by mak-dunkelziffer** [comment]
I generally prefer the pragmatic approach of having a good Rubocop rule enabled, but selectively turning it off where necessary. For example, we sometimes define inline classes in helpers and their instance variables are obviously fine. So I'd rather write:
```ruby
module MyHelper
  def my_helper
    # ivars bad
  end

  # rubocop:disable Rails/HelperInstanceVariable
  class MyInlineClass
    # ivars fine
  end
  # rubocop:enable Rails/HelperInstanceVariable
end
```
... than having no checks at all.

But I agree that turning this rule back on now will be really annoying for existing projects, so I'm fine with not doing it.

I also see, why Rubocop might be annoying, if using `rubocop:disable` comments is frowned upon.
